### ISSUE TITLE
react-router-redux@next. SSR state differs from client. Extra "key" field

### ISSUE THREAD

**Issue by dpostolachi** [remix-run/react-router#6055]
I have a problem when using React 16's hydrate on an app with SSR.

I get this warning when I try to get the state from SSR and use it as initial state in client's browser:

Warning: Prop `dangerouslySetInnerHTML` did not match. Server: "window.__PRELOADED_STATE__ = {\"routing\":{\"location\":{\"pathname\":\"\\u002Flist\",\"search\":\"?query=some\",\"hash\":\"\",\"key\":\"73omsy\"}}}" Client: "window.__PRELOADED_STATE__ = {\"routing\":{\"location\":{\"pathname\":\"\\u002Flist\",\"search\":\"?query=some\",\"hash\":\"\"}}}"

SSR:
```javascript
{
   routing:{
      location:{
         hash: "",
         key: "73omsy",
         pathname: "/list",
         search: "?query=some",
      }
   }
}
```

Client:
```javascript
{
   routing:{
      location:{
         hash: "",
         pathname: "/list",
         search: "?query=some",
      }
   }
}
```

Basically  the only problem is that location on server has an extra "key" field that doesn't get generated on client with history's browserHistory.

On Server I use memoryHistory and ConnectedRouter with isSRR, I push into history the url from request. If I delete the "key" field on Server it seems to work fine.

Is there another way to get the same state on the client and server?

**Comment by pshrmn** [comment]
@voodoo-child Are you not using the `<StaticRouter>` on the server? Unless I'm mistaken, the location generated by the server should not have a `key` property.

**Comment by voodoo-child** [comment]
@pshrmn No, I'm using ConnectedRouter, I've seen this isSSR prop for ConnectedRouter  and I thought there is no need to use two different routers anymore.

**Comment by pshrmn** [comment]
Ah, yes, sorry. That is what I get for not actually reading your whole post. I'm not actually familiar with `r-r-r`'s internals, so I can't be of much help here. The hack solution would be to delete the `key` property from the initial `history` location before rendering on the server.

```js
const history = createMemoryHistory();
delete history.location.key;
```

A better solution would probably require an update to the `history` package to re-use a provided `key` when initializing a browser history. I believe that adding a `key` to the initial location has been brought up before, but rejected, so I'm not sure if that will happen.

**Comment by voodoo-child** [comment]
Yeah, that's what I ended up doing. Thanks anyway.
